### PR TITLE
Update Ruby version to 3.0.6 (latest of 3.0 series)

### PR DESCRIPTION
**Pull request by heliocola**
## What type of PR is this? (check all applicable)
- [x] Ruby update

## Description
In preparation to try out Ruby 3.1.2 with YJIT, this PR udpate to latest version of the Ruby 3.0 series.
The main goal of this PR is to update to the latest 3.0 series and create an issue to discuss an strategy and benchmark for Ruby 3.1.2 with and without YJIT.

## Related Tickets & Documents

This PR replaces #19523 

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

No extra QA would be necessary, but let CI run would be ideal.

### UI accessibility checklist
None

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [x] No, and this is why: no extra tests needed.

### CODE REVIEW

File: .env_sample [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Development options
 
 # App core values
-APP_DOMAIN="localhost:3000"
+APP_DOMAIN="forem.test"
```
Comment (heliocola) [inline .env_sample:4]: On GH Actions, some tests are still failing with invalid domain.
```
1st Try error in ./spec/system/feedback_message_spec.rb:13:
invalid domain: ".localhost:3000"

RSpec::Retry: 2nd try ./spec/system/feedback_message_spec.rb:13

2nd Try error in ./spec/system/feedback_message_spec.rb:13:
invalid domain: ".localhost:3000"
RSpec::Retry: 3rd try ./spec/system/feedback_message_spec.rb:13
```

Comment (heliocola) [reply .env_sample:4]: Reference: https://github.com/forem/forem/actions/runs/5239228362/jobs/9458925719?pr=19603

**Comment by heliocola** [comment]
Forem Team: I reviewed all the CI failures and it looks like the issues with `invalid domain` have been fixed.
In fact, the change in .env_sample to not use `localhost:3000` is required.
References:
* https://github.com/forem/forem/pull/19603/files#r1226072028
* https://github.com/forem/forem/actions/runs/5239228362/jobs/9458925719?pr=19603

The current CI failures are grouped as follow:
1. Knapsack pro connection failures (both rspec and cypress)
2. And `quay.io/forem/ruby:3.0.6: not found`

My main goal with this PR is to be able to get to latest version of Ruby 3.0 series and from there update to Ruby 3.1.2 (and do some benchmarking with YJIT).
So, let me know if there is anything else needed to merge this.

And at last but not least: there is an issue with cookies domain, as noted by @gareth here: https://github.com/forem/forem/pull/19523#issuecomment-1582894095

That can probably be a separate issue to make Forem to conform with [RFC 6265](https://www.rfc-editor.org/rfc/rfc6265#section-4.1.1).

Despite the fact Rails is changing its behavior (ref: https://github.com/rails/rails/pull/48036 and thank you @gareth ), Forem repo does specifically adds leading dot to the domain in a few places, so that should be treated separately.
References:
* https://github.com/forem/forem/blob/main/app/lib/middlewares/set_cookie_domain.rb#L10
* https://github.com/forem/forem/blob/main/config/initializers/devise_monkeypatch.rb#L13
* https://github.com/forem/forem/blob/main/config/initializers/devise_monkeypatch.rb#L20

So that deserves its own issue, pr, and review process.

**Comment by mirie** [comment]
@maestromac flagging you on this one

**Comment by klardotsh** [comment]
Upgrading the `quay.io/forem/ruby` image is non-trivial unfortunately, on two axes:

- The image definition isn't currently open-sourced (there's not a super strong reason for this, but splitting it out of our in-house repo is still work that needs to be accounted for and time-boxed to a degree)
- The Ruby version installed in the container is just *whatever `dnf` pulls down from the Fedora repos at the time of image build*. Per Repology, [no version of Fedora packages 3.0.6](https://repology.org/project/ruby/versions); the closest we can get is 3.0.5 on Fedora 35 (which is the version this internal image currently targets). Building a newer Ruby from source (or installing via [ruby-build](https://github.com/rbenv/ruby-build), or whatever) is both possible and something I've dealt with before on prior org's infra teams, but it's again non-trivial, and work that needs to be planned out time-wise.
- Even if we solve all of that, the internal image fails to build due to some cryptic interaction between DNF, Docker, and (at least) my host filesystem:

```
 > [2/2] RUN dnf upgrade -y     && dnf install [REDACTED LIST OF PACKAGES]:
#0 78.68 Fedora 35 - x86_64                              1.0 MB/s |  79 MB     01:17    
#0 78.70 Error: Cannot create directory "/var/cache/dnf/fedora-37d3cae0527b6391/repodata": Invalid argument
```

This may be a weird "I use ZFS on my host system" quirk, or it may be an issue with Void Linux's Docker builds, or it may be an outright problem with how we build images. I'd have to do some testing (and probably tap @maestromac, who is on Fedora with, I presume, ext4 or btrfs root fs, for more data points to narrow down the cause of failure)

I suspect getting this PR merged is thus a minimum of a week of active (at 75%+ of a full-time engineer's available working time) effort.

File: .env_sample (hunk 1 above)
Comment (heliocola) [reply .env_sample:4]: @maestromac , @klardotsh : I just found in `bin/setup` that .env_sample is used as base for Dev/Local environment.
```
  unless File.exist?(".env")
    FileUtils.cp ".env_sample", ".env"
  end
```

So, I need to find a different solution for test environment.
I can simply have a .env.test with a different APP_CONFIG or change `environments/test.rb` to not consider this.

I will do some more digging on this tomorrow.
If you have a preferred way to go, please let me know.

**Comment by klardotsh** [comment]
Attempting to upgrade directly to Ruby 3.1+ does not solve point three above where the images currently do not build (at least for me, so far). That's something we need to figure out internally before we can take *any* upgrades to Ruby at all.

I do think taking the final patch release of the 3.0 series before moving to 3.1 is a good idea in general, so I want to keep this PR going and advocate internally (I've already started these discussions) for making the time to unblock it in the ways mentioned above :)

**Comment by klardotsh** [comment]
Additionally: for what it's worth, the `forem/ruby` image isn't only necessary for CI purposes, it's also what eventually underlies `forem/forem:development`, which (if I recall correctly) is built out of the `Dockerfile` at the root of the repo, and is what powers all `docker-compose` based workflows - for at least some of us internally, and I'm sure others in the open-source community, that is our primary way of running Forem locally for development, and can't be ignored or sidelined.

File: .env_sample (hunk 1 above)
Comment (jaw6) [reply .env_sample:4]: Hi @heliocola — I'm not sure about this change (I'll defer to @maestromac and @klardotsh there) but the Ruby upgrade looks good without it, and this change seems unrelated. I'd suggest maybe moving this change to a separate PR, that way the Ruby upgrade can proceed on its own?

File: Gemfile [hunk 2]
```diff
@@ -173,3 +173,5 @@ group :test do
   gem "with_model", "~> 2.1.6" # Dynamically build a model within an RSpec context
   gem "zonebie", "~> 0.6.1" # Runs your tests in a random timezone
 end
+
+gem "cgi", "~> 0.3.6" # Support for the Common Gateway Interface protocol.
```
Comment (jaw6) [inline Gemfile:177]: For other reviewers:

* 3.0.5 addresses [a response-splitting vulnerability in the CGI gem](https://www.ruby-lang.org/en/news/2022/11/22/http-response-splitting-in-cgi-cve-2021-33621/) and encourages pinning it to a version > `0.3.5`
* Also, 3.0.3 addresses [a buffer overrun in the CGI gem](https://www.ruby-lang.org/en/news/2021/11/24/buffer-overrun-in-cgi-escape_html-cve-2021-41816/) 

We haven't upgraded our Ruby in [over a year](https://www.ruby-lang.org/en/news/2021/07/07/ruby-3-0-2-released/) 😱

**Review by jaw6 — CHANGES_REQUESTED** [review]
@heliocola I really appreciate you opening this PR! I'm mostly a :+1: but as mentioned, I'd prefer to split the proposed `.env` change into a separate PR so we can merge this one without delay.

**Comment by klardotsh** [comment]
Hey @heliocola, just letting you know that the infra team here at Forem (that's @maestromac and I) are digging into this actively. I don't yet know when we'll have this unblocked, but hopefully within the coming days to weeks. Steps we're taking:

- Working on a new `forem/ruby` container image where Ruby versions will track the upstream [`_/ruby` containers](https://hub.docker.com/_/ruby/), rather than tracking the Ruby versions Fedora packages.
- This new container image will be open-source as part of this `forem/forem` repo, rather than living in our internal company infrastructure repo (which is private).
- I'm going to be personally testing ARM64 support (this refs #16765) as part of this: I'm making absolutely no promises, it's kind of a tangential/stretch goal that @benhalpern and I talked about and agreed was worth seeing if we could address as part of the same work.
- Automating the build and publish steps of that base image so that future updates should be relatively straightforward (bumping a version number in the base Dockerfile and opening a PR).

I'll circle back when I know more. Thanks also to @jaw6 for the review above: I agree that I'd prefer to see `.env` changes separated to a different PR that can be reviewed on its own merits. I don't actually know that `forem.test` is the right value for `APP_DOMAIN`, given that there are no `A` records in public domain registrars for it (is this something I was supposed to add to `/etc/hosts` and just never did?):

```
(woods) ~  » drill forem.test
;; ->>HEADER<<- opcode: QUERY, rcode: NXDOMAIN, id: 10163
;; flags: qr aa rd ra ; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0 
;; QUESTION SECTION:
;; forem.test.	IN	A

;; ANSWER SECTION:

;; AUTHORITY SECTION:
.	86400	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2023062202 1800 900 604800 86400

;; ADDITIONAL SECTION:

;; Query time: 24 msec
;; SERVER: 1.1.1.1
;; WHEN: Thu Jun 22 12:21:08 2023
;; MSG SIZE  rcvd: 103
```

**Comment by klardotsh** [comment]
Addendum: reffing https://github.com/forem/forem/issues/19626 as well here (the issue form of the discussion I linked above).

File: .env_sample (hunk 1 above)
Comment (heliocola) [reply .env_sample:4]: I think splitting the .env for testing is the best way to go.
I will push that change and sync up this PR.

@maestromac : for development environment, this variable isn't used for cookie the way it is for Testing and Production.
Check these references: https://github.com/forem/forem/blob/main/config/initializers/middlewares.rb#L4C89-L4C89
But it still doesn't explain how this fails in the testing environment.

I will split the env file, and do some digging if I remember how the test env is affected.

**Comment by heliocola** [comment]
@maestromac , @klardotsh , @jaw6 : TL'DR: I will close this PR and open a fresh one & clean one.
Once this is unblocked, it can be reviewed again, and merged.

## The Long story

Here is a summary of my latest changes:
### 1. Removed the sym link on .env.test and create its own file.
I've tried to change `COMMUNITY_NAME` to be `DEV(test)` but I found out rspec and cypress tests expecting for `DEV(local)` so I've changed back the value in .env.test to `DEV(local)`.
I did however refactored some rspec tests to check for what is in `.env.test`.

### 2. Fixed/Changed some rubocop offenses
Duo to changing some files for the item above, I've had to fix some rubocop offenses.
The one that seems most important is:
```ruby
  click_button("button", text: /\ASave changes\z/)
```
instead of
```ruby
  find("button", text: /\ASave changes\z/).click
```
But this change cause CI because click_button only works when button is disabled and serveral test were failing with:
```
Capybara::ElementNotFound:
       Unable to find button "button" that is not disabled
```
But reverting it, I have to wrestle with rubocop to commit the changes.
So, given this PR is stale and blocked, I will close and open a fresh one with the changes needed for Ruby 3.0.6.

**Pull request closed without merging**